dbms/src/Storages/DeltaMerge/Index/LocalIndexBuilder.h

```diff
@@ -0,0 +1,53 @@
+// Copyright 2024 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <Columns/ColumnVector.h>
+#include <Columns/IColumn.h>
+#include <DataTypes/IDataType.h>
+#include <Storages/DeltaMerge/Index/LocalIndexInfo.h>
+#include <Storages/DeltaMerge/Index/LocalIndex_fwd.h>
+
+
+namespace DB::DM
+{
+
+/// Builds a LocalIndex.
+class LocalIndexBuilder
+{
+public:
+    using ProceedCheckFn = std::function<bool()>;
+
+public:
+    static LocalIndexBuilderPtr create(const LocalIndexInfo & index_info);
+
+public:
+    explicit LocalIndexBuilder(const LocalIndexInfo & index_info)
+        : index_info(index_info)
+    {}
+
+    virtual ~LocalIndexBuilder() = default;
+
+    virtual void addBlock(const IColumn & column, const ColumnVector<UInt8> * del_mark, ProceedCheckFn should_proceed)
```

Comment:
Honestly I prefer to not extract and have this encapsulation at this moment, because in FTS we may want to support multiple columns (e.g. a FULLTEXT index over `(title, body)`). It will completely change the interface. We can unify interface after most of FTS is stabilized.

Reply:
Then how to use different builder in `LocalIndexWriter`?

Reply:
Using the current way (`switch kind`)? I mean this encapausulation is likely to be changed again so maybe we can defer this change to later.